src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/OffenderService.kt

```diff
@@ -28,40 +28,65 @@ class OffenderService(
-            logAndTrackEvent(PARTIAL_MATCH_MESSAGE, TelemetryEventType.DELIUS_PARTIAL_MATCH_FOUND, personEntity, person)
-          }
+  private fun getOffenderMatcher(person: Person): OffenderMatcher {
+    val offenderDetails = client.getOffenderDetail(SearchDto.from(person))
+    return OffenderMatcher(offenderDetails, person)
+  }
 
-          multipleOffendersMatch(offenderDetails, person) -> {
-            offenderDetails.forEach { offender ->
-              logAndTrackEvent(MULTIPLE_MATCHES_MESSAGE, TelemetryEventType.DELIUS_MATCH_FOUND, personEntity, person)
-              personRecordService.addOffenderToPerson(personEntity, Person.from(offender))
-            }
-          }
-        }
-      }
-    }
+  private fun noRecordsFound(personEntity: PersonEntity, person: Person) {
+    logAndTrackEvent(NO_RECORDS_MESSAGE, TelemetryEventType.DELIUS_NO_MATCH_FOUND, personEntity, person)
+  }
+
+  private fun partialMatchFound(personEntity: PersonEntity, person: Person) {
+    logAndTrackEvent(PARTIAL_MATCH_MESSAGE, TelemetryEventType.DELIUS_PARTIAL_MATCH_FOUND, personEntity, person)
   }
 
-  private fun isMatchingOffender(person: Person, offenderDetail: OffenderDetail) =
-    person.otherIdentifiers?.pncIdentifier == PNCIdentifier(offenderDetail.otherIds.pncNumber) &&
-      offenderDetail.firstName.equals(person.givenName, true) &&
-      offenderDetail.surname.equals(person.familyName, true) &&
-      offenderDetail.dateOfBirth == person.dateOfBirth
+  private fun exactMatchFound(offenderMatches: OffenderMatcher, personEntity: PersonEntity, person: Person) {
+    logAndTrackEvent(EXACT_MATCH_MESSAGE, TelemetryEventType.DELIUS_MATCH_FOUND, personEntity, person)
+    val offenderDetail = offenderMatches.getOffenderDetail()
+    personRecordService.addOffenderToPerson(personEntity, Person.from(offenderDetail))
+  }
+
+  private fun multipleMatchesFound(offenderMatches: OffenderMatcher, personEntity: PersonEntity, person: Person) {
+    val allMatchingOffenders = offenderMatches.getAllMatchingOffenders()
+    allMatchingOffenders?.forEach { offenderDetail ->
+      logAndTrackEvent(MULTIPLE_MATCHES_MESSAGE, TelemetryEventType.DELIUS_MATCH_FOUND, personEntity, person)
+      personRecordService.addOffenderToPerson(personEntity, Person.from(offenderDetail))
+    }
+  }
 
-  @Suppress("UNCHECKED_CAST")
+  private fun trackPncMismatchEvent(
+    offenderDetails: List<OffenderDetail>,
+    personEntity: PersonEntity,
+    person: Person,
+  ) {
+    telemetryService.trackEvent(
+      TelemetryEventType.DELIUS_PNC_MISMATCH,
+      mapOf(
+        "UUID" to personEntity.personId.toString(),
+        "PNC searched for" to person.otherIdentifiers?.pncIdentifier?.pncId,
+        "PNC returned from search" to offenderDetails.joinToString(" ") { it.otherIds.pncNumber.toString() },
+        "PRISON NUMBER" to offenderDetails.singleOrNull()?.otherIds?.nomsNumber,
+      ).filterValues { !it.isNullOrBlank() },
+    )
+  }
```

Comment:
I think I can merge this with the other one. let me try something.

Reply:
Merging makes the code unnecessarily messy by passing extra parameters without the need for it.